sstables/storage.cc

```diff
@@ -36,10 +36,57 @@
 
 namespace sstables {
 
+class dir_wrapper final {
+    std::filesystem::path _pathname;
+    file _file;
+    bool _file_stale;
+
+public:
+    explicit dir_wrapper(std::filesystem::path pathname) : _pathname(std::move(pathname)) {};
+    dir_wrapper(const dir_wrapper&) = delete;
+    dir_wrapper& operator=(const dir_wrapper&) = delete;
+    dir_wrapper(dir_wrapper&&) = delete;
+    dir_wrapper& operator=(dir_wrapper&&) = delete;
+    ~dir_wrapper() = default;
+
+    const std::filesystem::path::string_type& native() const noexcept {
+        return _pathname.native();
+    }
+
+    operator std::filesystem::path() const noexcept {
+        return _pathname;
+    }
+
+    dir_wrapper& operator=(const sstring &nd) {
+        _pathname = nd;
+        _file_stale = true;
+        return *this;
+    }
+
+    future<> sync(const sstable& sst) {
+        if (_file && _file_stale) {
+            co_await sst.sstable_write_io_check(std::mem_fn(&file::close), _file);
+            _file = file();
+        }
+        if (!_file) {
+            _file = co_await sst.sstable_write_io_check(open_directory, _pathname.native());
+            _file_stale = false;
+        }
+        co_await sst.sstable_write_io_check(std::mem_fn(&file::flush), _file);
+    };
+
+    future<> destroy(const sstable& sst) {
+        if (_file) {
+            co_await sst.sstable_write_io_check(std::mem_fn(&file::close), _file);
+            _file = file();
```

Comment:
I'm not sure you need this assignment.
a `file` should evaluate to `false` after `close()` since it will have no file impl.

Reply:
I added the assignment because of the following comment in `seastar/include/seastar/core/file.hh` (at 908ccd936a63):
```
    /// One can check whether a file object is in uninitialized state with
    /// \ref operator bool(); One can reset a file back to uninitialized state
    /// by assigning file() to it.
```

The corresponding comment on `close()` does not say that the file object is reset to uninitialized state.

I do agree `file::close()` in `seastar/src/core/file.cc` empties `_file_impl` by moving it.

Reply:
We can rely on it, and send a PR to update the documentation. file::close is terminal in the sense that nothing can be done with file after ot except destroying it (and we also allow closing multiple times, but I don't think we need to guarantee and document that).
So we can safely document that after close, a file returns to the uninitialized state.

Reply:
> So we can safely document that after close, a file returns to the uninitialized state.

https://github.com/scylladb/seastar/pull/2353

Reply:
.. also removed the assignment (both instances), as advised.

(The entire enclosing blocks might disappear in v5, once I manage to introduce `_close_fut` for replacing `_file_stale`; I've not progressed that far yet.)

Reply:
https://github.com/scylladb/seastar/pull/2353 has been merged meanwhile, and the async close is part of v6; resolving this thread.

Reply:
(thanks for reviewing that seastar PR btw)